// @roots/bud-client/lib/components/indicator/indicator.controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Controller = void 0;
/**
 * Activity indicator controller
 * @public
 */
class Controller {
    /**
     * Initialization
     * @public
     */
    constructor() {
        /**
         * Active WHM payload
         * @public
         */
        this.payload = null;
        this.node = document.createElement(`bud-activity-indicator`);
        this.update = this.update.bind(this);
    }
    /**
     * Append `bud-error` element to the DOM
     *
     * @public
     */
    addNode() {
        var _a;
        if (document.body.querySelector(`bud-activity-indicator`)) {
            if (typeof this.timer.unref === `function`)
                this.timer.unref();
            this.removeNode();
        }
        (_a = document.body) === null || _a === void 0 ? void 0 : _a.appendChild(this.node);
        this.timer = setTimeout(this.removeNode, 3000);
    }
    /**
     * Remove `bud-error` element from the DOM (if present)
     *
     * @public
     */
    removeNode() {
        var _a;
        (_a = document.body.querySelector(`bud-activity-indicator`)) === null || _a === void 0 ? void 0 : _a.remove();
    }
    /**
     * Update activity indicator
     * @public
     */
    update(payload) {
        var _a, _b;
        this.node.toggleAttribute(`has-errors`, ((_a = payload.errors) === null || _a === void 0 ? void 0 : _a.length) ? true : false);
        this.node.toggleAttribute(`has-warnings`, ((_b = payload.warnings) === null || _b === void 0 ? void 0 : _b.length) ? true : false);
        this.node.setAttribute(`action`, payload.action);
        this.addNode();
    }
}
exports.Controller = Controller;
