// @roots/bud-build/lib/item/item.js
import { __decorate } from "tslib";
import { bind } from '@roots/bud-support/decorators';
import { isString } from '@roots/bud-support/lodash-es';
import { basename } from 'path';
import Loader from '../loader/loader.js';
import Base from '../shared/base.js';
/**
 * Item class
 *
 * @public
 */
class Item extends Base {
    /**
     * Class constructor
     *
     * @public
     */
    constructor(_app, constructorParams) {
        super(_app);
        this._app = _app;
        constructorParams?.ident && this.setIdent(constructorParams.ident);
        constructorParams?.loader && this.setLoader(constructorParams.loader);
        !constructorParams?.ident &&
            constructorParams?.loader &&
            this.setIdent(isString(constructorParams.loader)
                ? constructorParams.loader
                : basename(constructorParams.loader.getSrc()));
        constructorParams?.options &&
            this.setOptions(constructorParams.options);
    }
    getIdent() {
        return this.ident;
    }
    setIdent(ident) {
        this.ident = ident;
        return this;
    }
    /**
     * Get rule set item loader
     *
     * @public
     * @decorator `@bind`
     */
    getLoader() {
        return this.loader instanceof Loader
            ? this.loader
            : this.app.build.loaders[this.loader];
    }
    /**
     * Set rule set item loader
     *
     * @public
     * @decorator `@bind`
     */
    setLoader(loader) {
        this.loader = loader;
        if (!this.ident)
            this.setIdent(basename(isString(loader) ? loader : loader.getSrc()));
        return this;
    }
    /**
     * Get rule set item options
     *
     * @public
     * @decorator `@bind`
     */
    getOptions() {
        return this.unwrap(this.options);
    }
    /**
     * Set rule set item options
     *
     * @public
     * @decorator `@bind`
     */
    setOptions(options) {
        this.options = options;
        return this;
    }
    /**
     * Merge rule set item options
     *
     * @public
     * @decorator `@bind`
     */
    mergeOptions(options) {
        this.setOptions({
            ...(this.getOptions() ?? {}),
            ...options,
        });
        return this;
    }
    /**
     * Produce rule set item object for Webpack
     *
     * @public
     * @decorator `@bind`
     */
    toWebpack() {
        const output = {
            loader: this.getLoader()?.getSrc(),
        };
        if (this.options) {
            output.options = this.getOptions();
        }
        if (this.ident) {
            output.ident = this.getIdent();
        }
        if (!output.loader) {
            this.app.error(`error in ${this.ident}`, `no loader registered`);
        }
        return output;
    }
}
__decorate([
    bind
], Item.prototype, "getIdent", null);
__decorate([
    bind
], Item.prototype, "setIdent", null);
__decorate([
    bind
], Item.prototype, "getLoader", null);
__decorate([
    bind
], Item.prototype, "setLoader", null);
__decorate([
    bind
], Item.prototype, "getOptions", null);
__decorate([
    bind
], Item.prototype, "setOptions", null);
__decorate([
    bind
], Item.prototype, "mergeOptions", null);
__decorate([
    bind
], Item.prototype, "toWebpack", null);
export { Item as default };
