// @roots/bud-client/lib/index.js
"use strict";
/* eslint-disable no-console */
/* global __resourceQuery */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
require("./interface");
const components = __importStar(require("./components/index.js"));
const events_js_1 = require("./hmr/events.js");
const clientOptions = __importStar(require("./options"));
((queryString) => __awaiter(void 0, void 0, void 0, function* () {
    /**
     * Webpack hot interface
     */
    const webpackHot = module.hot;
    /* Set client options from URL params */
    const options = clientOptions.setFromParameters(queryString);
    /**
     * Returns true if environment supports HMR
     */
    if (typeof (window === null || window === void 0 ? void 0 : window.EventSource) === `undefined`) {
        console.error(`[bud] hot module reload requires EventSource to work. https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events#Tools`);
        return false;
    }
    /**
     * Webpack HMR check handler
     */
    const check = () => __awaiter(void 0, void 0, void 0, function* () {
        return webpackHot.status() === `idle` &&
            (yield webpackHot.check(false).then((modules) => __awaiter(void 0, void 0, void 0, function* () {
                modules && (yield update());
            })));
    });
    /**
     * Webpack HMR unaccepted module handler
     */
    const onUnacceptedOrDeclined = (info) => {
        console.warn(`[${options.name}] ${info.type}`, info);
        options.reload && window.location.reload();
    };
    /**
     * Webpack HMR error handler
     */
    const onErrored = error => {
        var _a, _b;
        const message = `[${(_a = error === null || error === void 0 ? void 0 : error.moduleId) !== null && _a !== void 0 ? _a : options.name}] ${(_b = error === null || error === void 0 ? void 0 : error.error) !== null && _b !== void 0 ? _b : `error`}`;
        console.error(message);
        components.controllers.map(controller => controller.update({
            type: `accept-errored`,
            action: `built`,
            errors: [{ message }],
        }));
    };
    /**
     * Webpack HMR update handler
     */
    const update = () => __awaiter(void 0, void 0, void 0, function* () {
        webpackHot.status() === `ready` &&
            (yield webpackHot.apply({
                autoApply: false,
                ignoreUnaccepted: true,
                ignoreDeclined: true,
                ignoreErrored: true,
                onErrored,
                onUnaccepted: onUnacceptedOrDeclined,
                onDeclined: onUnacceptedOrDeclined,
            }));
    });
    /* Instantiate eventSource */
    const eventSource = events_js_1.Events.make(options);
    /* Instantiate indicator, overlay */
    yield components.make(options);
    /* Instantiate HMR event source and register client listeners */
    eventSource.addMessageListener((event) => __awaiter(void 0, void 0, void 0, function* () {
        try {
            const data = JSON.parse(event.data);
            if (!data)
                return;
            if (data.action === `reload`)
                window.location.reload();
            components.controllers.map(controller => controller.update(data));
            if (__webpack_hash__ !== eventSource.currentHash)
                yield check();
        }
        catch (error) { }
    }));
}))(__resourceQuery);
