// @roots/bud-client/lib/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setFromParameters = exports.get = exports.data = void 0;
/**
 * Client options
 * @public
 */
let data = {
    [`default`]: {
        timeout: 20 * 1000,
        reload: true,
        name: `default`,
        debug: true,
        log: true,
        path: `/__bud/hmr`,
        indicator: true,
        overlay: true,
    },
};
exports.data = data;
/**
 * Get client option
 * @public
 */
const get = (name, key) => key ? data[name][key] : data[name];
exports.get = get;
/**
 * Set client data based on URL parameters
 */
const setFromParameters = (query) => {
    let parsedParams = {};
    new window.URLSearchParams(query).forEach((value, key) => {
        parsedParams[key] =
            value === `true` ? true : value === `false` ? false : value;
    });
    data[parsedParams.name] = Object.assign(Object.assign({}, data.default), parsedParams);
    return data[parsedParams.name];
};
exports.setFromParameters = setFromParameters;
