// @roots/bud-server/lib/middleware/hot/middleware.js
/* eslint-disable no-console */
import { HotEventStream } from './stream.js';
const middlewarePath = `/__bud/hmr`;
let latestStats = null;
let closed = false;
let logger;
export default (app) => {
    logger = app.logger.instance.scope(...app.logger.scope, `hmr`);
    return makeHandler(app.compiler.instance);
};
export const makeHandler = (compiler) => {
    const stream = new HotEventStream();
    const onInvalid = () => {
        if (closed)
            return;
        latestStats = null;
        stream.publish({ action: `building` });
    };
    const onDone = (stats) => {
        if (closed)
            return;
        latestStats = stats;
        publish(`built`, latestStats, stream);
    };
    compiler.hooks.invalid.tap(`bud-hot-middleware`, onInvalid);
    compiler.hooks.done.tap(`bud-hot-middleware`, onDone);
    const middleware = function (req, res, next) {
        if (closed)
            return next();
        if (!req.url.endsWith(middlewarePath))
            return next();
        stream.handle(req, res);
        if (latestStats) {
            publish(`sync`, latestStats, stream);
        }
    };
    // @ts-ignore
    middleware.publish = function (payload) {
        if (closed)
            return;
        stream.publish(payload);
    };
    // @ts-ignore
    middleware.close = function () {
        if (closed)
            return;
        closed = true;
        stream.close();
    };
    return middleware;
};
export const publish = (action, statsCompilation, stream) => {
    const bundles = collectCompilations(statsCompilation.toJson({
        all: false,
        cached: true,
        children: true,
        modules: true,
        timings: true,
        hash: true,
        errors: true,
    }));
    bundles.forEach((stats) => {
        const name = stats.name ?? statsCompilation.name ?? `unnamed`;
        const modules = stats.modules?.reduce((modules, module) => ({ ...modules, [module.id]: module.name }), {});
        logger.log(`built`, name, `(${stats.hash})`, `in`, `${stats.time}ms`);
        stream.publish({
            name,
            action,
            time: stats.time,
            hash: stats.hash,
            warnings: stats.warnings || [],
            errors: stats.errors || [],
            modules,
        });
    });
};
export const collectCompilations = (stats) => {
    // Stats has modules, single bundle
    if (stats.modules)
        return [stats];
    // Stats has children, multiple bundles
    if (stats.children && stats.children.length)
        return stats.children;
    // Not sure, assume single
    return [stats];
};
