// @roots/bud/lib/services/env/index.js
import { __decorate } from "tslib";
import { Service } from '@roots/bud-framework/service';
import { bind } from '@roots/bud-support/decorators';
import Container from '@roots/container';
/**
 * Env service
 *
 * @public
 */
export default class Env extends Service {
    /**
     * Class constructor
     *
     * @public
     */
    constructor(app) {
        super(app);
        this.data = new Container();
        this.has = this.data.has;
        this.get = this.data.get;
        this.set = this.data.set;
        this.isFunction = this.data.isFunction;
        this.isString = this.data.isString;
    }
    /**
     * Bootstrap event callback
     *
     * @public
     * @decorator `@bind`
     */
    async bootstrap() {
        this.data.setStore(this.app.context.env);
    }
    /**
     * Get entries from .env which include `APP_PUBLIC` in their name
     *
     * @public
     * @decorator `@bind`
     * @decorator `@once`
     */
    getPublicEnv() {
        return this.data
            .getEntries()
            .filter(this.filterPublicEnv)
            .map(this.transformPublicEnv)
            .reduce((a, [k, v]) => ({ ...a, [k]: v }), {});
    }
    /**
     * Transform public env
     *
     * @public
     * @decorator `@bind`
     */
    transformPublicEnv([rawKey, rawValue]) {
        return [rawKey.replace(`PUBLIC_`, ``), rawValue];
    }
    /**
     * Filter public env
     *
     * @public
     */
    filterPublicEnv([key]) {
        return key.startsWith(`PUBLIC_`);
    }
}
/**
 * Service label
 *
 * @public
 */
Env.label = `env`;
__decorate([
    bind
], Env.prototype, "bootstrap", null);
__decorate([
    bind
], Env.prototype, "getPublicEnv", null);
__decorate([
    bind
], Env.prototype, "transformPublicEnv", null);
__decorate([
    bind
], Env.prototype, "filterPublicEnv", null);
