// @roots/bud-build/lib/handlers/loaders.js
import MiniCssExtractPlugin from 'mini-css-extract-plugin';
/**
 * css loader
 *
 * @public
 */
export const css = (app) => {
    const loader = `css-loader`;
    return app.build.makeLoader().setSrc(loader);
};
/**
 * csv loader
 *
 * @public
 */
export const csv = (app) => {
    const loader = `csv-loader`;
    return app.build.makeLoader().setSrc(loader);
};
/**
 * file loader
 *
 * @public
 */
export const file = (app) => {
    const loader = `file-loader`;
    return app.build.makeLoader().setSrc(loader);
};
/**
 * html-loader
 *
 * @public
 */
export const html = (app) => {
    const loader = `html-loader`;
    return app.build.makeLoader().setSrc(loader);
};
/**
 * remark-loader
 *
 * @public
 */
export const remark = (app) => {
    const loader = `remark-loader`;
    return app.build.makeLoader().setSrc(loader);
};
/**
 * `mini-css-extract-plugin.loader`
 *
 * @public
 */
export const minicss = (app) => {
    return app.build.makeLoader().setSrc(MiniCssExtractPlugin.loader);
};
/**
 * style-loader
 *
 * @public
 */
export const style = (app) => {
    const loader = `style-loader`;
    return app.build.makeLoader().setSrc(loader);
};
/**
 * xml-loader
 *
 * @public
 */
export const xml = (app) => {
    const loader = `xml-loader`;
    return app.build.makeLoader().setSrc(loader);
};
/**
 * yml-loader
 *
 * @public
 */
export const yml = (app) => {
    const loader = `yml-loader`;
    return app.build.makeLoader().setSrc(loader);
};
